### PR TITLE
nfs: add support for snapshot, restore and pvc-clone

### PR DESCRIPTION
**Pull request by Rakshith-R**
- 1d7191080 e2e: add tests for nfs snapshot and restore
- 3d6b57949 e2e: add pvc-pvc clone e2e test
- c44e83383 deploy: add nfs pvc-clone & pod-clone example yaml
- 827b14b8d nfs: add support for pvc-pvc clone
- f0b28b794 deploy: add support for nfs snapshot
- d7e4b910d nfs: add support for create & delete snapshot

### CODE REVIEW

**Review by nixpanic — CHANGES_REQUESTED** [review]
looks good to me, just need to update the README that includes the support for the new features

**Review by yati1998 — COMMENTED** [review]
LGTM

File: deploy/nfs/kubernetes/csi-provisioner-rbac.yaml [hunk 1]
```diff
@@ -39,6 +39,21 @@ rules:
...
+    verbs: ["create", "get", "list", "watch", "update", "delete", "patch"]
+  - apiGroups: ["snapshot.storage.k8s.io"]
+    resources: ["volumesnapshotclasses"]
+    verbs: ["get", "list", "watch"]
+  - apiGroups: ["snapshot.storage.k8s.io"]
+    resources: ["volumesnapshotcontents/status"]
+    verbs: ["update", "patch"]
+  - apiGroups: ["snapshot.storage.k8s.io"]
+    resources: ["volumesnapshots/status"]
+    verbs: ["update", "patch"]
```
Comment (Madhu-1) [inline deploy/nfs/kubernetes/csi-provisioner-rbac.yaml]: Please reverify once do we need all these access?

File: e2e/nfs.go [hunk 2]
```diff
@@ -569,6 +583,346 @@ var _ = Describe("nfs", func() {
...
+
+				app.Namespace = f.UniqueName
+				app.Spec.Volumes[0].PersistentVolumeClaim.ClaimName = pvc.Name
+				label := make(map[string]string)
+				label[appKey] = appLabel
+				app.Labels = label
+				opt := metav1.ListOptions{
+					LabelSelector: fmt.Sprintf("%s=%s", appKey, label[appKey]),
+				}
+				wErr := writeDataInPod(app, &opt, f)
```
Comment (Madhu-1) [inline e2e/nfs.go]: use err here?

File: e2e/nfs.go [hunk 3]
```diff
@@ -569,6 +583,346 @@ var _ = Describe("nfs", func() {
...
+				if failed != 0 {
+					e2elog.Failf("deleting PVCs and apps failed, %d errors were logged", failed)
+				}
+
+				parentPVCCount := totalSubvolumes - totalCount
+				validateSubvolumeCount(f, parentPVCCount, fileSystemName, subvolumegroup)
+				validateOmapCount(f, parentPVCCount, cephfsType, metadataPool, volumesType)
+				validateOmapCount(f, totalCount, cephfsType, metadataPool, snapsType)
+				// create clones from different snapshots and bind it to an
+				// app
```
Comment (Madhu-1) [inline e2e/nfs.go]: single-line comment should be fine?

File: e2e/nfs.go [hunk 4]
```diff
@@ -569,6 +583,346 @@ var _ = Describe("nfs", func() {
...
+				label[appKey] = appLabel
+				app.Labels = label
+				opt := metav1.ListOptions{
+					LabelSelector: fmt.Sprintf("%s=%s", appKey, label[appKey]),
+				}
+				wErr := writeDataInPod(app, &opt, f)
+				if wErr != nil {
+					e2elog.Failf("failed to write data from application %v", wErr)
+				}
+
```
Comment (Madhu-1) [inline e2e/nfs.go]: writing data is required? if yes please add md5 checksum to make sure the content is matching.

File: e2e/nfs.go (hunk 2 above)
Comment (Madhu-1) [inline e2e/nfs.go]: writing data is required? if yes please add md5 checksum to make sure the content is matching.

File: e2e/snapshot.go [hunk 5]
```diff
@@ -206,6 +206,58 @@ func createCephFSSnapshotClass(f *framework.Framework) error {
...
+	}
+	sc.Parameters["clusterID"] = fsID
+	sclient, err := newSnapshotClient()
+	if err != nil {
+		return err
+	}
+	_, err = sclient.VolumeSnapshotClasses().Create(context.TODO(), &sc, metav1.CreateOptions{})
+	if err != nil {
+		return fmt.Errorf("failed to create volumesnapshotclass: %w", err)
+	}
```
Comment (Madhu-1) [inline e2e/snapshot.go]: extra create is not required as we are already creating in the PollImmediate loop?

File: e2e/snapshot.go [hunk 6]
```diff
@@ -206,6 +206,58 @@ func createCephFSSnapshotClass(f *framework.Framework) error {
...
+func deleteNFSSnapshotClass() error {
+	scPath := fmt.Sprintf("%s/%s", nfsExamplePath, "snapshotclass.yaml")
+	sc := getSnapshotClass(scPath)
+
+	sclient, err := newSnapshotClient()
+	if err != nil {
+		return err
+	}
+
+	return sclient.VolumeSnapshotClasses().Delete(context.TODO(), sc.Name, metav1.DeleteOptions{})
```
Comment (Madhu-1) [inline e2e/snapshot.go]: use same retry mechanism to delete also?

**Review by Madhu-1 — COMMENTED** [review]
no major comments, some nits. LGTM, as @nixpanic mentioned Doc update is required

File: deploy/nfs/kubernetes/csi-provisioner-rbac.yaml (hunk 1 above)
Comment (Rakshith-R) [reply deploy/nfs/kubernetes/csi-provisioner-rbac.yaml]: > Please reverify once do we need all these access?

okay, I am copying these permissions
https://github.com/kubernetes-csi/external-snapshotter/blob/v5.0.1/deploy/kubernetes/csi-snapshotter/rbac-csi-snapshotter.yaml#L33-L41

**Comment by Rakshith-R** [comment]
@mergifyio rebase

File: README.md [hunk 7]
```diff
@@ -114,6 +114,9 @@ for its support details.
 |        | Dynamically provision, de-provision File mode ROX volume  | Alpha          | >= v3.6.0          | >= v1.0.0        | Pacific (>=16.2.0)   | >= v1.14.0         |
 |        | Dynamically provision, de-provision File mode RWOP volume | Alpha          | >= v3.6.0          | >= v1.5.0        | Pacific (>=16.2.0)   | >= v1.22.0         |
 |        | Expand volume                                             | Alpha          | >= v3.7.0          | >= v1.1.0        | Pacific (>=16.2.0)   | >= v1.15.0         |
+|        | Creating and deleting snapshot                            | Alpha          | >= v3.7.0          | >= v1.1.0        | Pacific (>=16.2.0)   | >= v1.15.0         |
+|        | Provision volume from snapshot                            | Alpha          | >= v3.7.0          | >= v1.1.0        | Pacific (>=16.2.0)   | >= v1.15.0         |
+|        | Provision volume from another volume                      | Alpha          | >= v3.7.0          | >= v1.1.0        | Pacific (>=16.2.0)   | >= v1.15.0         |
```
Comment (Madhu-1) [inline README.md]: reverify kubernetes support version it should be 1.17 for snapshot and 1.16 for PVC cone

**Review by Madhu-1 — COMMENTED** [review]
changes LGTM. small clarification is required.

File: README.md (hunk 7 above)
Comment (Rakshith-R) [reply README.md]: > reverify kubernetes support version it should be 1.17 for snapshot and 1.16 for PVC cone

missed it, using the correct version now, ptal

**Review by nixpanic — DISMISSED** [review]

**Comment by nixpanic** [comment]
Both CI jobs failed with RBD image count out-of-sync

**Comment by nixpanic** [comment]
**ci/centos/mini-e2e/k8s-1.22** failed with the new tests:
```
create a PVC clone and bind it to an app
...
May 18 11:52:43.119: INFO: waiting for PVC nfs-80811 (600 seconds elapsed)
May 18 11:52:43.121: INFO: failed to create PVC and app (nfs-80810): timed out waiting for the condition
May 18 11:52:43.121: INFO: failed to create PVC and app (nfs-80811): timed out waiting for the condition
May 18 11:52:43.121: INFO: failed to create PVC and app (nfs-80812): timed out waiting for the condition
May 18 11:52:43.121: FAIL: creating PVCs and apps failed, 3 errors were logged
```
https://jenkins-ceph-csi.apps.ocp.ci.centos.org/blue/organizations/jenkins/mini-e2e_k8s-1.22/detail/mini-e2e_k8s-1.22/3076/pipeline

**Comment by Rakshith-R** [comment]
> **ci/centos/mini-e2e/k8s-1.22** failed with the new tests:
> 
> ```
> create a PVC clone and bind it to an app
> ...
> May 18 11:52:43.119: INFO: waiting for PVC nfs-80811 (600 seconds elapsed)
> May 18 11:52:43.121: INFO: failed to create PVC and app (nfs-80810): timed out waiting for the condition
> May 18 11:52:43.121: INFO: failed to create PVC and app (nfs-80811): timed out waiting for the condition
> May 18 11:52:43.121: INFO: failed to create PVC and app (nfs-80812): timed out waiting for the condition
> May 18 11:52:43.121: FAIL: creating PVCs and apps failed, 3 errors were logged
> ```
> 
> https://jenkins-ceph-csi.apps.ocp.ci.centos.org/blue/organizations/jenkins/mini-e2e_k8s-1.22/detail/mini-e2e_k8s-1.22/3076/pipeline

```
May 18 11:52:43.138: INFO: At 2022-05-18 11:42:43 +0100 BST - event for nfs-80811: {persistentvolume-controller } ExternalProvisioning: waiting for a volume to be created, either by external provisioner "nfs.csi.ceph.com" or manually created by system administrator
May 18 11:52:43.138: INFO: At 2022-05-18 11:42:43 +0100 BST - event for nfs-80812: {nfs.csi.ceph.com_csi-nfsplugin-provisioner-b45fbb776-k7sm7_c5a32210-c1aa-42bd-a3de-6a4ef61f2e74 } ProvisioningFailed: failed to provision volume with StorageClass "csi-nfs-sc": error getting handle for DataSource Type VolumeSnapshot by Name nfs-80810: error getting snapshot nfs-80810 from api server: volumesnapshots.snapshot.storage.k8s.io "nfs-80810" is forbidden: User "system:serviceaccount:cephcsi-e2e-2a199dae:nfs-csi-provisioner" cannot get resource "volumesnapshots" in API group "snapshot.storage.k8s.io" in the namespace "nfs-8081"

```

external provisioner needs to access to volumesnapshot get/list to restore pvc from snapshot

**Review by Madhu-1 — DISMISSED** [review]

**Review by yati1998 — DISMISSED** [review]

**Comment by Rakshith-R** [comment]
The mountPath did'nt match from pod.yaml to pod-clone.yaml and pod-restore.yaml
fixed it 
https://github.com/ceph/ceph-csi/compare/810832a49fbe149d281c8b8bff376b580ed421c0..fdfb0fe6a00f51fea2b73a45ecff43e87258f0b3
PTAL

File: deploy/nfs/kubernetes/csi-provisioner-rbac.yaml [hunk 8]
```diff
@@ -39,6 +39,18 @@ rules:
   - apiGroups: ["storage.k8s.io"]
     resources: ["csinodes"]
     verbs: ["get", "list", "watch"]
+  - apiGroups: ["snapshot.storage.k8s.io"]
+    resources: ["volumesnapshotclasses"]
+    verbs: ["get", "list", "watch"]
+  - apiGroups: ["snapshot.storage.k8s.io"]
+    resources: ["volumesnapshotcontents"]
+    verbs: ["create", "get", "list", "watch", "update", "delete", "patch"]
```
Comment (Madhu-1) [inline deploy/nfs/kubernetes/csi-provisioner-rbac.yaml]: create and delete access is not required remove it https://github.com/kubernetes-csi/external-snapshotter/blob/master/deploy/kubernetes/csi-snapshotter/rbac-csi-snapshotter.yaml#L37-L38?

Comment (Rakshith-R) [reply deploy/nfs/kubernetes/csi-provisioner-rbac.yaml]: > create and delete access is not required remove it https://github.com/kubernetes-csi/external-snapshotter/blob/master/deploy/kubernetes/csi-snapshotter/rbac-csi-snapshotter.yaml#L37-L38?

release v5.0.1 which we are still using has this rbac, 
Let's update when we get v6.0.0 release

**Comment by Rakshith-R** [comment]
https://github.com/ceph/ceph-csi/compare/2738cf9ac4d78591316d159b577dfa24e8b47ebf..a5a3e2a42afe1b39fc0a7f0996be07e68d7bd389

only parent PVC had to be deleted. 
demo pod had been deleted during call to 
https://github.com/ceph/ceph-csi/blob/7e1ed75cfbc0ba4465b1945c4663f95934f773ac/e2e/utils.go#L772

**Comment by Rakshith-R** [comment]
:partying_face: ci is all green, PTAL

https://jenkins-ceph-csi.apps.ocp.ci.centos.org/blue/organizations/jenkins/mini-e2e_k8s-1.23/detail/mini-e2e_k8s-1.23/1748/pipeline/89
https://jenkins-ceph-csi.apps.ocp.ci.centos.org/blue/organizations/jenkins/mini-e2e_k8s-1.22/detail/mini-e2e_k8s-1.22/3114/pipeline/89

**Review by Madhu-1 — APPROVED** [review]

**Comment by Rakshith-R** [comment]
/retest ci/centos/mini-e2e-helm/k8s-1.22
ci passed but
jenkins is not updating the status :disappointed: 
https://jenkins-ceph-csi.apps.ocp.ci.centos.org/blue/organizations/jenkins/mini-e2e-helm_k8s-1.22/detail/mini-e2e-helm_k8s-1.22/3225/pipeline

**Comment by Rakshith-R** [comment]
@nixpanic @yati1998 can you please take a look?

**Comment by Rakshith-R** [comment]
@mergifyio rebase

**Review by nixpanic — APPROVED** [review]

**Pull request merged**